refactor: [M3-7376] - Region Select Autocomplete

abailly-akamai opened the pull request:
## Description 📝
This PR's primary goal is to refactor the RegionSelect to use the new Autocomplete component.
It is a precursor to the work that needs to be done for DC Get Well. This refactor will help facilitating the addition of logic and API fetching within the RegionSelect.

I apologize for the diff size. It touches a lot of files that may not seem related, albeit benefited from the rework. It was hard to compartementalize some of the changes that felt natural to be a part of the PR. Please comment and ask for details as needed.

## Changes  🔄
- Replace the enhanced select with Autocomplete and configure to get parity
- Extract styles, types and utils for better separation of concerns
- Add relevant tests for extracted utils
- Replace all instanced of `selectedID` with `selectedId`
- Remove redundancy with Flag types typed as lowercase 
- Add a new e2e dedicated to the behavior of RegionSelect
- Update all relevant e2e tests
- Update all relevant unit tests

## Preview 📷
There should not be any significant visual change to the RegionSelect (maybe very minor things) and the behavior should be identical. 

**Note**: the only (intentional) change you should see is the ordering of the regions. The region ordering and DCs ordering is sorted alphabetically, except for "North America" which comes first (same as current RegionSelect)

## How to test 🧪

### Prerequisites
- Pull and run code locally

### Verification steps 
Verify the behavior and appearance of all all the places the RegionSelect is implemented (create flows, migration, clone, One click apps, Stackscripts etc)

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [x] ♿  Providing accessibility support

abailly-akamai commented on packages/manager/cypress/e2e/core/images/machine-image-upload.spec.ts:
```diff
@@ -120,7 +126,7 @@ const uploadImage = (label: string) => {
   getClick('[id="label"][data-testid="textfield-input"]').type(label);
   getClick('[id="description"]').type('This is a machine image upload test');
-  fbtClick('Select a Region');
+  containsPlaceholderClick(selectRegionString);
```
This will be the preferred way to trigger the select from now on (new `containsPlaceholderClick` util + using the constant rather than a hard coded string. 

This has been replaced in all suites CC @jdamore-linode

abailly-akamai commented on packages/manager/cypress/e2e/core/linodes/create-linode.spec.ts:
```diff
@@ -30,12 +31,95 @@ import {
+    // Confirm that selecting a valid region updates the Plan Selection panel.
+    expect(cy.get('[data-testid="table-row-empty"]').should('not.exist'));
+  });
+
```
This is a bit of a "Component" test I stuck in the Linode Create flow. It will help us make sure we got some extra coverage and a dedicated unit for the RegionSelect when working on DC Get Well. CC @jdamore-linode

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,150 @@
+      />
+    </Box>
+  );
+});
```
The core of the refactor. Meant to be a 1/1 replacement for this iteration.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.utils.ts:
```diff
@@ -0,0 +1,125 @@
+) => {
+  const allRegions = [
+    ...regions,
+    ...listOfFakeRegions
```
Reminder fake regions are going away with DC Get Well, but this PR is a 1/1 replacement of the current RegionSelect so the logic still belongs here

abailly-akamai commented on packages/manager/src/components/RegionSelect/disabledRegions.tsx:
```diff
@@ -0,0 +1,103 @@
+/* Those fake regions are a way to display a region that is not available for selection, and removed from the API response.
+/* They are a visual indicator that the region is not available, and a link to learn more about why, as to not confuse the user when they don't see the region in the list.
+/* The whole concept is going away with DC Get Well (which is great news cause this isn't a sustainable feature), since the API will provide a new endpoint to evaluate the DC's availability.
+*/
```
See comment above

abailly-akamai commented on packages/manager/src/dev-tools/FeatureFlagTool.tsx:
```diff
@@ -21,7 +21,6 @@ const options: { flag: keyof Flags; label: string }[] = [
   { flag: 'dcSpecificPricing', label: 'DC-Specific Pricing' },
   { flag: 'objDcSpecificPricing', label: 'OBJ Storage DC-Specific Pricing' },
   { flag: 'selfServeBetas', label: 'Self Serve Betas' },
-  { flag: 'soldOutTokyo', label: 'Sold Out Tokyo' },
```
I removed this from the menu settings, it's not needed anymore

abailly-akamai commented on packages/manager/src/featureFlags.ts:
```diff
@@ -61,6 +61,7 @@ export interface Flags {
   referralBannerText: ReferralBannerText;
   regionDropdown: boolean;
   selfServeBetas: boolean;
+  soldOutSingapore: boolean;
```
Adding `soldOutSingapore` in the off chance we need to turn it on before DC Get Well is released. Since the fake menu item is controlled by a feature flag we can close the [hotfix](https://github.com/linode/manager/pull/9805) and we won't need another release if we need to turn it on. 

This will be cleaned up with DC Get Well

abailly-akamai commented on packages/manager/src/features/Databases/DatabaseCreate/DatabaseCreate.tsx:
```diff
@@ -494,7 +494,7 @@ const DatabaseCreate = () => {
               engineOptions
             )}
             className={classes.engineSelect}
-            components={{ Option: RegionOption, SingleValue: _SingleValue }}
```
kinda weird to use this here so I created a new EngineOption to separate concerns. No change, just a new dumb component rather than using the deprecated RegionOption

abailly-akamai commented on packages/manager/src/features/Databases/DatabaseCreate/EngineOption.tsx:
```diff
@@ -0,0 +1,52 @@
+      </Grid>
+    </Option>
+  );
+};
```
See above

abailly-akamai commented on packages/manager/src/features/Linodes/MigrateLinode/ConfigureForm.tsx:
```diff
@@ -128,20 +128,13 @@ export const ConfigureForm = React.memo((props: Props) => {
             handleSelection={handleSelectRegion}
             label="New Region"
-            menuPlacement="top"
-            selectedID={selectedRegion}
```
We don't need those styles with the new autocomplete. Those were here to address some inconsistencies with EnhancedSelect

bnussman-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,150 @@
+
+  return (
+    <Box sx={{ width }}>
+      <Autocomplete
```
I don't see `getOptionDisabled` being used with the AutoComplete. Does it apply here?

bnussman-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,150 @@
+        errorText={errorText}
+        groupBy={(option: RegionSelectOption) => option.data.region}
+        label={label ?? 'Region'}
+        options={options}
```
Would there be any benefit to just passing `regions` directly and doing any extra computation in `renderOption`?

bnussman-akamai replied:
On second thought, we might be blocked by the "faked" regions on this

abailly-akamai replied:
It will be cleaned up real soon I promise 👍

bnussman-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,150 @@
+    handleSelection(selection?.value);
+  };
+
+  const options: any = React.useMemo(
```
`any` can be removed I think

bnussman-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,150 @@
+  const regionFromSelectedId =
+    getSelectedRegionById(regions, selectedId ?? '') ?? null;
+
+  const [
+    selectedRegion,
+    setSelectedRegion,
+  ] = React.useState<RegionSelectOption | null>(regionFromSelectedId);
```
Do we need to maintain state in this component?

abailly-akamai replied:
It does not bother me too much right now, and it helps with the clone flow, but I'll keep this in mind when implementing DC Get Well

bnussman-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
Can we resolve the `any` in `RegionSelectOption`?

coliu-akamai reviewed (commented):
Great work Alban! 🎉 

✅ region select did function and look as expected for the flows
✅ unit tests
edit: 
✅ most cypress tests were working for me!
❓ Looking into running the cypress tests in the /regions package --found pr 9372 so will try that

abailly-akamai commented:
@coliu-akamai as mentioned during standup I was refactoring some e2e utils and tests to align with the new conventions. I just pushed those fixes now so tests should succeed nicely now 👍

coliu-akamai commented:
Awesome thanks!

coliu-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,150 @@
+    handleSelection(selection?.value);
+  };
+
+  const options: any = React.useMemo(
```
duplicate comment, resolving

coliu-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.types.ts:
```diff
@@ -0,0 +1,27 @@
+import type { EnhancedAutocompleteProps } from 'src/components/Autocomplete/Autocomplete';
+
+export interface RegionSelectOption {
+  data?: any;
```
duplicate comment, resolving

abailly-akamai commented on packages/manager/src/features/Databases/DatabaseCreate/EngineOption.tsx:
```diff
@@ -0,0 +1,52 @@
+
+import type { OptionProps } from 'react-select';
+
+const useStyles = makeStyles()((theme: Theme) => ({
+  focused: {
+    backgroundColor: theme.palette.primary.main,
+    color: 'white',
+  },
+  root: {
+    padding: theme.spacing(1),
+  },
+}));
```
I don't think it matters much either way so I'll leave this as be for now.

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.styles.ts:
```diff
@@ -0,0 +1,43 @@
+
+import { ListItem } from 'src/components/ListItem';
+
+export const GroupHeader = styled('div')(({ theme }) => ({
```
done!

abailly-akamai commented on packages/manager/cypress/e2e/core/linodes/create-linode.spec.ts:
```diff
@@ -142,6 +228,8 @@ describe('create linode', () => {
     const initialRegion = getRegionById('us-west');
     const newRegion = getRegionById('us-east');
 
+    console.log(initialRegion);
```
noooope. ty!

bnussman-akamai commented on packages/manager/src/__data__/regionsData.ts:
```diff
@@ -8,47 +8,47 @@ const resolvers = {
   {
     capabilities: ['Linodes', 'NodeBalancers', 'Block Storage'],
-    country: 'in',
+    country: 'IN',
```
Countries returned by the API are all lowercase. Can we reflect that in the type?

abailly-akamai replied:
Ah yeah, my bad - i cleaned this up and reverted the data changes

bnussman-akamai commented on packages/manager/cypress/e2e/core/images/create-linode-from-image.spec.ts:
```diff
@@ -55,12 +55,19 @@ const createLinodeWithImageMock = (url: string, preselectedImage: boolean) => {
+authenticate();
 describe('create linode from image, mocked data', () => {
+  before(() => {
+    cleanUp(['linodes']);
```
Is this needed if we're only using mock data?

bnussman-akamai commented on packages/manager/cypress/e2e/core/linodes/clone-linode.spec.ts:
```diff
@@ -126,7 +124,7 @@ describe('clone linode', () => {
    * - Confirms that notice is shown when selecting a region with a different price structure.
    */
-  it('shows DC-specific pricing information during clone flow', () => {
+  it.only('shows DC-specific pricing information during clone flow', () => {
```
Can this `only` be removed?

abailly-akamai replied:
yesssss - thx for catching

bnussman-akamai reviewed (commented):
Functionality is looking good. Should we try to tighten up the spacing a bit?

| develop | this PR |
|--------|-------|
|   ![Screenshot 2023-11-08 at 10 05 59 AM](https://github.com/linode/manager/assets/115251059/945b250d-3959-459c-8fd3-47a70f81a348)     |    ![Screenshot 2023-11-08 at 10 05 55 AM](https://github.com/linode/manager/assets/115251059/a87448b0-a16c-4497-920b-844cdc01be0d)   |

jdamore-linode commented on packages/manager/cypress/support/ui/autocomplete.ts:
```diff
@@ -0,0 +1,73 @@
+    return autocompletePopper.findByTitle(`${region.label} (${region.id})`);
+  },
+
+  /**
+   * Finds and open the region select input.
+   */
+  open: (): Cypress.Chainable => {
+    return cy.get('[data-testid="region-select"] input').click();
+  },
```
Would it be possible to limit the UI helpers to selection and keep interaction within the tests?

e.g. `ui.regionSelect.find().click()`. 

Sticking to this convention gives us more flexibility to interact with the components in the test --  for example, a test author may want to find the region select and assert it's visible without interacting with it, or may want to interact with the region select by typing rather than by clicking it open (and in this case there's a little bit of ambiguity about what interaction is actually happening since you can open the region select menu in a couple ways).

I don't want to commit to defining every possible interaction for a UI element in the UI helpers, but I also don't want there to be some subset of interactions that are handled by the UI helpers and leave it to test authors to have to figure out what is and what isn't there -- I think it's most straightforward to leave it to the test.

abailly-akamai replied:
Yeah that makes perfect sense - I'll update those

abailly-akamai replied:
All set! also fixed the x-axis clickaway behavior

jdamore-linode commented:
I'm seeing a minor issue where I can't dismiss the autocomplete popper if I click anywhere within the element containing the `data-qa-autocomplete` attribute

https://github.com/linode/manager/assets/97627410/40ded3c6-bac2-4214-8fbe-e5ceac0c87ba

jaalah-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,161 @@
+                } as ListItemComponentsPropsOverrides,
+              }}
+            >
+              {Boolean(option.data.disabledMessage) ? (
```
What do you think about removing the ternary is just doing 
```
<Box alignItems="center" display="flex" flexGrow={1}>
  <StyledFlagContainer>
    <Flag country={option.data.country} />
  </StyledFlagContainer>
  {option.label}
  {Boolean(option.data.disabledMessage) && ' (Not available)'}
</Box>
{selected && <SelectedIcon visible={selected} />}
```

abailly-akamai replied:
much cleaner 👍 - done!

jaalah-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,161 @@
+          <Tooltip
+            disableFocusListener={Boolean(!option.data.disabledMessage)}
+            disableHoverListener={Boolean(!option.data.disabledMessage)}
+            disableTouchListener={Boolean(!option.data.disabledMessage)}
```
Nit: Should we create a const for these?

abailly-akamai replied:
sure thing, done

jaalah-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -0,0 +1,161 @@
+        textFieldProps={{
+          InputProps: {
+            required,
+            startAdornment: selectedRegion && (
```
Nice! 🔥

jaalah-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.utils.ts:
```diff
@@ -0,0 +1,127 @@
+  ];
+
+  return allRegions
+    .map((region: Region) => {
```
This is a great place to add a return type since we have one `RegionSelectOption`

abailly-akamai replied:
yup, done as well!

jaalah-akamai reviewed (commented):
Excellent work @abailly-akamai - just a few comments

coliu-akamai reviewed (approved):
🎉 Thanks Alban!
Re-reviewed functionality + tests and code ✅

jdamore-linode commented on packages/manager/cypress/e2e/core/linodes/create-linode.spec.ts:
```diff
@@ -57,20 +140,17 @@ describe('create linode', () => {
   });
 
-  it('creates a linode via CLI', () => {
+  it.only('creates a linode via CLI', () => {
```
```suggestion
  it('creates a linode via CLI', () => {
```

jdamore-linode commented on packages/manager/cypress/e2e/core/linodes/create-linode.spec.ts:
```diff
@@ -30,12 +29,95 @@ import {
+
+    cy.get('[data-qa-option="eu-west"]').click();
+    cy.get('[data-qa-autocomplete-popper="true"]').should('not.exist');
+    autocomplete.within(() => {
+      cy.get('[data-testid="textfield-input"]').should(
+        'have.value',
+        'London, UK (eu-west)'
+      );
+    });
```
```suggestion
    ui.regionSelect.find().within(() => {
      cy.get('[data-testid="textfield-input"]').should(
        'have.value',
        'London, UK (eu-west)'
      );
    });
// (And line 82 can be removed.)
```
This is an [antipattern](https://docs.cypress.io/guides/references/best-practices#Assigning-Return-Values) in Cypress and can cause unexpected test behavior.

If you want to refer to an element that was selected before, you'll want to do it [with an alias](https://docs.cypress.io/guides/core-concepts/variables-and-aliases#Sharing-Context) instead of a const assignment, for example:

```typescript
ui.regionSelect.find().as('autocomplete');

// Later...
cy.get('@autocomplete').within(() => { ... })
```

(I think, essentially, if the page or component re-renders after you assign the const, it'll fail when you try to use it later in the test)

jdamore-linode commented on packages/manager/cypress/e2e/core/linodes/create-linode.spec.ts:
```diff
@@ -30,12 +29,95 @@ import {
+      });
+
+    // Confirm that region select dropdown is populated with expected regions, sorted alphabetically.
+    cy.get('[data-qa-option]')
+      .should('have.length', 5)
+      .should((item) => {
+        expect(item[0]).to.contain('Dallas, TX (us-central)');
+        expect(item[1]).to.contain('Newark, NJ (us-east)');
+        expect(item[2]).to.contain('Singapore, SG (ap-south)');
+        expect(item[3]).to.contain('Tokyo, JP (ap-northeast) (Not available)');
+        expect(item[4]).to.contain('London, UK (eu-west)');
+      });
```
I'm getting a test failure here:

```
Expected undefined to have a length of 5 but got 0
```
I'll try to follow up with a fix or a better explanation in a few

abailly-akamai commented:
@jdamore-linode thanks for catching

- removed the .only offender
- fixed the tests that were tripping and used better assertions
- added a backlog item to lint for .only which will be useful for e2e and unit tests

jdamore-linode reviewed (approved):
Looks great! Thanks @abailly-akamai!
